### DISCUSSION TITLE
Get MethodInfo from managed function pointer?

### DISCUSSION THREAD

**Discussion by acaly** [dotnet/runtime#46215] [Q&A] [answered]
C# 9.0 starts to support function pointers of managed methods in unsafe context. Although it's designed to be performance-orientated, I think it would still be nice to have an API to find the ```MethodInfo``` from the pointer in some cases, for example, to allow the program to provide debug info or to confirm that the signature of the function pointer is really correct.

I am not exactly sure whether this is easy or even possible in .NET Core, but I think the ```StackTrace``` class should be providing similar functionality internally. So is it possible to expose such functions in the core library?

**Comment by PathogenDavid** [comment, marked as answer]
If you don't strictly need a `MethodInfo`, but rather just information about the method the function pointer refers to, you can use [ClrMD](https://github.com/microsoft/clrmd) for this.

For example:

```csharp
using System;
using System.Diagnostics;
using System.Runtime.InteropServices;
using Microsoft.Diagnostics.Runtime;
using System.Linq;

public static unsafe class Program
{
    public static void Main(string[] args)
    {
        IntPtr user32 = NativeLibrary.Load("User32.dll");

        delegate*<void> managed1 = &TestMethod;
        delegate*<void> managed2 = &AnotherMethod;
        delegate*<void> managed3 = &MessageBoxW;
        delegate* unmanaged<void> unmanaged1 = &UnmanagedMethod;
        delegate* unmanaged<void> unmanaged2 = (delegate* unmanaged<void>)NativeLibrary.GetExport(user32, "MessageBoxW");

        // Note that this means any methods JITted after we create the snapshot won't be available
        // You can use AttachToProcess on yourself, but it's not supported.
        // https://github.com/microsoft/clrmd/blob/master/doc/FAQ.md#can-i-use-this-api-to-inspect-my-own-process
        using DataTarget target = DataTarget.CreateSnapshotAndAttach(Process.GetCurrentProcess().Id);
        ClrRuntime runtime = target.ClrVersions.First().CreateRuntime();

        void Test(string testName, void* functionPointer)
        {
            ClrMethod? method = runtime.GetMethodByInstructionPointer((ulong)functionPointer);

            if (method is null)
            {
                Console.WriteLine($"{testName}: Not found");
                return;
            }

            Console.WriteLine($"{testName}: {method.Signature}");
        }

        Test("managed1", managed1);
        Test("managed2", managed2);
        Test("managed3", managed3);
        Test("unmanaged1", unmanaged1);
        Test("unmanaged2", unmanaged2); // This is expected to not be found because it's a native method
    }

    public static void TestMethod()
    { }

    public static void AnotherMethod()
    {}

    [UnmanagedCallersOnly]
    public static void UnmanagedMethod()
    { }

    [DllImport("User32.dll")]
    public static extern void MessageBoxW(); // Obviously this won't be callable, it's just here so it resolves
}
```

Here's the output you can expect from this:

```
managed1: Program.TestMethod()
managed2: Program.AnotherMethod()
managed3: Program.MessageBoxW()
unmanaged1: Program.UnmanagedMethod()
unmanaged2: Not found
```

(`unmanaged2` isn't found because it's a function pointer to native code.)

---------------

If you *need* a `MethodInfo`, it seems to be possible using the information provided by ClrMD, but from what I can find there's no public API to get a `MethodInfo` back from an raw method handle. There is however an internal API for it: [`RuntimeType.GetMethodBase`](https://github.com/dotnet/runtime/blob/fed418c134b2b6359dc39ef8d312422f5918be0e/src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs#L1744).

I made a nasty little proof of concept to call it using reflection just to prove it works:

```csharp
using System;
using System.Reflection;

internal static class MethodBaseHelper
{
    private static Type? RuntimeMethodHandleInternal;
    private static ConstructorInfo? RuntimeMethodHandleInternal_Constructor;
    private static Type? RuntimeType;
    private static MethodInfo? RuntimeType_GetMethodBase;

    public static MethodBase? GetMethodBaseFromHandle(IntPtr handle)
    {
        RuntimeMethodHandleInternal ??= typeof(RuntimeMethodHandle).Assembly.GetType("System.RuntimeMethodHandleInternal", throwOnError: true)!;
        RuntimeMethodHandleInternal_Constructor ??= RuntimeMethodHandleInternal.GetConstructor
        (
            BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DoNotWrapExceptions,
            binder: null,
            new[] { typeof(IntPtr) },
            modifiers: null
        ) ?? throw new InvalidOperationException("RuntimeMethodHandleInternal constructor is missing!");

        RuntimeType ??= typeof(Type).Assembly.GetType("System.RuntimeType", throwOnError: true)!;
        RuntimeType_GetMethodBase ??= RuntimeType.GetMethod
        (
            "GetMethodBase",
            BindingFlags.NonPublic | BindingFlags.Static | BindingFlags.DoNotWrapExceptions,
            binder: null,
            new[] { RuntimeType, RuntimeMethodHandleInternal },
            modifiers: null
        ) ?? throw new InvalidOperationException("RuntimeType.GetMethodBase is missing!");

        // Wrap the handle
        object runtimeHandle = RuntimeMethodHandleInternal_Constructor.Invoke(new[] { (object)handle });
        return (MethodBase?)RuntimeType_GetMethodBase.Invoke(null, new[] { null, runtimeHandle });
    }
}
```

If you add the following after the `Console.WriteLine` in the `Test` local function:

```csharp
            MethodBase? methodBase = MethodBaseHelper.GetMethodBaseFromHandle((IntPtr)method.MethodDesc);
            Console.WriteLine($"    MethodBase: {methodBase?.Name ?? "Not Found"}");
```

You get the following output:

```
managed1: Program.TestMethod()
    MethodBase: TestMethod
managed2: Program.AnotherMethod()
    MethodBase: AnotherMethod
managed3: Program.MessageBoxW()
    MethodBase: MessageBoxW
unmanaged1: Program.UnmanagedMethod()
    MethodBase: UnmanagedMethod
unmanaged2: Not found
```

If you only want this for debugging then maybe this is good enough. If not, you'd probably have to open an API request in [dotnet/runtime](https://github.com/dotnet/runtime). Not sure how they'll feel about an API to get a `MethodInfo` from a function pointer directly, but I feel like safe, public wrapper to get a `MethodBase` back from a raw method handle `IntPtr` isn't all that unreasonable. (I'm kinda surprised it doesn't already exist, actually.)

If you want to avoid using ClrMD, you could probably loop over all methods using reflection and compare using [`RuntimeMethodHandle.GetFunctionPointer`](https://docs.microsoft.com/en-us/dotnet/api/system.runtimemethodhandle.getfunctionpointer), but that'd be super slow (unless you narrowed things down to just a specific type or assembly) and you'd be running into dangerous territory with tiered JITting (and other undescribed issues): https://github.com/dotnet/roslyn/issues/48919

**Reply by PathogenDavid** [reply]
I should note I tested this using ClrMD 2.0.161401 (which is the current latest.)

There's nothing special going on in it, but here's my full csproj:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    <Nullable>enable</Nullable>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Diagnostics.Runtime" Version="2.0.161401" />
  </ItemGroup>
</Project>

```

**Reply by acaly** [reply]
Thanks for the answer! Using CLRmd is a nice solution, but I am afraid it won't be scalable, as it creates a dump for the entire process every time a single function pointer is inspected (if I understand correctly, it suspends the CLR and call OS API to do that). I would prefer using internal API as a temporary solution.

As a permanent solution I really hope there could be a new public API. Maybe it should go into the ```Marshal``` class? Do you have any suggestions?

**Reply by PathogenDavid** [reply]
That is true, ClrMD might chug pretty hard creating a snapshot if your process is consuming a lot of memory. You could always use an external process that you communicate with over IPC, but then you might run into not wanting to suspend everything whenever ClrMD is doing its thing. Really depends on your specific needs.

(Heck, depending on your needs you could just connect ClrMD to your own process without suspending. It's officially not supported, but it's worked for me in the past.)

> I would prefer using internal API as a temporary solution.

To clarify, the internal API is complementing ClrMD in this case, not replacing it. I'm not aware of any accessible (internal or otherwise) API to get a method handle from a function pointer outside of ClrMD.

> As a permanent solution I really hope there could be a new public API. Maybe it should go into the Marshal class? Do you have any suggestions?

My instinct is it should go on `RuntimeMethodHandle` to mirror the existing `RuntimeMethodHandle.GetFunctionPointer`. Maybe something like this:

```csharp
namespace System
{
    public struct RuntimeMethodHandle
    {
        public static bool TryGetFromFunctionPointer(IntPtr functionPointer, out RuntimeMethodHandle handle);

        // ...existing members...
    }
}
````

Then you use [`MethodBase.GetMethodFromHandle`](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.methodbase.getmethodfromhandle) to get the actual `MethodInfo`.

**Reply by acaly** [reply]
I spent some time digging into ClrMD and it turns out to be using [```ISOSDacInterface::GetMethodDescPtrFromIP```](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/debugging/isosdacinterface-getmethoddescptrfromip-method). The ```ISOSDacInterface``` can be created with [```CLRDataCreateInstance```](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/debugging/clrdatacreateinstance-function), which requires a ```ICLRDataTarget``` object to be passed as argument. I think we only need to implement this interface in order to get CLR debugging info for the running process. [This file](https://github.com/dotnet/runtime/blob/69e114c1abf91241a0eeecf1ecceab4711b8aa62/src/coreclr/utilcode/dacutil.cpp) in dotnet/runtime has an implementation, and it seems that it should work with only the ```GetImageBase``` (probably corresponding to ```GetModuleHandle```) and ```ReadVirtual``` (simply read a block of memory) methods implemented, which is reasonable because all debugging info are just in the process's virtual memory. If these are true, we might be able to implement a ```ICLRDataTarget``` interface and create the ```ISOSDacInterface``` interface to call ```GetMethodDescPtrFromIP```.

Though I doubt this is realistic either. We might come into multithread-racing problem as, I imagine, the EE is constantly updating its lists of compiled methods using its internal locks that we cannot obtain. Probably that's also the reason why ClrMD does not support directly inspecting the the process itself is running in.

**Reply by PathogenDavid** [reply]
> We might come into multithread-racing problem as, I imagine, the EE is constantly updating its lists of compiled methods using its internal locks that we cannot obtain. Probably that's also the reason why ClrMD does not support directly inspecting the the process itself is running in.

I would agree. These APIs are intended for debugging, and generally assume the process is suspended when they're accessed. Something like this as an official feature would likely not go through a debugging API but instead something internal to the runtime.

**Reply by reflectronic** [reply]
An API similar to the `TryGetFromFunctionPointer` idea was previously proposed here: https://github.com/dotnet/runtime/issues/36363

According to Jan,
> The reverse lookup would be slow, and supporting it adds extra constrains on the runtime to support a niche scenario. I do not see it as a good tradeoff. If you want to use reflection, stick with reflection objects, without roundtripping through function pointers.

**Reply by acaly** [reply]
@reflectronic  Thanks for the link!

If the maintainers have decided that it would be difficult to implement this, I don't have much hope for having this feature in any foreseeable future. I might go with a mixed strategy with reflection and function pointer.

**Reply by PathogenDavid** [reply]
@reflectronic Nice find, surprised neither of us stumbled upon that earlier.

------

Cc @Sergio0694 It's not practical for ComputeSharp's purposes, but you might find the above workaround of interest.

**Reply by PathogenDavid** [reply]
@danmosemsft @terrajobst This is a good real-world example of [my point that locking old API proposals is harmful](https://github.com/dotnet/runtime/issues/44775#issuecomment-730227417). There's no way to share the above ClrMD-based solution to those who might be subscribed to or stumble upon [the linked proposal](https://github.com/dotnet/runtime/issues/36363).

**Reply by GrabYourPitchforks** [reply]
In theory this can also be done using _public_ reflection. Create a managed delegate whose signature matches the function pointer signature. For instance, `System.Action` matches `delegate*<void>`. Every managed delegate type has a public `.ctor(object, IntPtr)`, where the first argument corresponds with _this_ (or _null_ if pointing to a static method) and the second argument corresponds to the managed function pointer target. You can't call this constructor directly from C#, but you can use public reflection to locate and invoke it.

Once you have the managed delegate instance created, you can query the public `Delegate.Method` property to retrieve the method info.

**Reply by PathogenDavid** [reply]
Ooh, that's much cleaner than the ClrMD workaround. Here's a modified version of my test from earlier: https://gist.github.com/PathogenDavid/7e860057edeb398c6b2fdb42535d666c

Only downside is it (unsurprisingly) violently explodes with an `ExecutionEngineException` if you pass it a native function pointer.

Interestingly, it doesn't actually seem to matter if the signature of the delegate matches the actual method. IE: If you change `Action` to `Action<int>` the delegate is still created with the correct `MethodInfo`. (Obviously invoking those delegates would result in pain.)

**Reply by PathogenDavid** [reply]
I briefly dug into the runtime and it looks like [`Entry2MethodDesc`](https://github.com/dotnet/runtime/blob/c64861b3bb0b3e31f70aabee875ded8cfd4dbaf7/src/coreclr/vm/method.cpp#L2278-L2320) (called from [`DelegateConstruct`](https://github.com/dotnet/runtime/blob/69e114c1abf91241a0eeecf1ecceab4711b8aa62/src/coreclr/vm/comdelegate.cpp#L1561-L1564)) is the main function of interest.

It might be worth re-requesting https://github.com/dotnet/runtime/issues/36363, it seems to me like it'd be easier to provide this functionality than originally thought.

**Reply by acaly** [reply]
I actually noticed this solution when I thought about this myself (because normal delegate creation is just done like this), but did not try it as I was worried if I don't get the correct type the runtime might fail in the ctor.

However, I just tried it and even if the target method's signature is wrong (e.g., using ```Action``` for a ```void Method(int x)```), the ```Delegate``` is still able to retrieve the ```MethodInfo``` from the runtime. So I think this probably is the best workaround we can achieve right now.

@PathogenDavid maybe you want to update your example with different signatures?

**EDIT** sorry I didn't notice your comment on the same question

**Reply by GrabYourPitchforks** [reply]
FWIW I don't think the runtime gives any sort of guarantee if you try to construct a delegate with an incompatible signature type, so I wouldn't rely on that behavior. But if you pass it a managed function pointer with a compatible signature, from the runtime's perspective it's quite literally the same behavior that the C# compiler would've produced when generating a normal managed delegate around that same method, so it should be dependable.

**Reply by acaly** [reply]
> I don't think the runtime gives any sort of guarantee if you try to construct a delegate with an incompatible signature type

That's probably a good reason to have this API. As I said, one usage of this API is to validate the signature. If we don't know about the signature, then we cannot construct the delegate. Also it's not always easy to have a delegate with the same signature. `Action` and `Func` cannot be used with byrefs. I suggest we reopen #36363.

**Reply by GrabYourPitchforks** [reply]
> Also it's not always easy to have a delegate with the same signature.

Not sure what you mean by this. If you have a `delegate* managed<..., TRet>`, it should always be possible to construct an equivalent `delegate TRet MyDelegate(...)`. I can't think offhand of any instances where this wouldn't be possible. The exception would be if you're literally passing the function pointer around as a `void*` or an `IntPtr` or equivalent, but it'd be somewhat useless to do that since you can't invoke it without knowing the exact signature.

**Reply by acaly** [reply]
> it'd be somewhat useless to do that since you can't invoke it without knowing the exact signature

That is true, unless you are writing something more generic (literally). Without the support of byref types as generic argument, and something like a variadic template in C++, one will find it difficult to write a library that helps user of the library to handle function pointers. I hope the first can be solved by the ref struct constraint proposal, and then the second can be somehow solved by defining a lot of generic types/methods with different number of generic parameters like `Action` and `Func`. 

Please consider my following situation. I am working on an ECS library for games. Briefly, the function pointers, created by plugins, which do know about the signature, go through the library, which has no information of the signature and generally can only store them as `IntPtr`. These function pointers can be consumed by another plugin, who again, knows about the signature. I cannot use `Delegate` for some reasons, so there would be no `Delegate` types to pass around. Instead, the consumer defines a helper struct (with significant help of a source generator) to store the info and invoke the method. You can think of this as a re-implementation of the C# delegate. The runtime needs `IntPtr` to implement `Delegate`s, so I would also have to use `IntPtr`.

I can let the producer to provide `MethodInfo` together with the function pointer in Debug builds for validation, or I can let the library to obtain the information from a `Action` delegate as described above.

**Reply by GrabYourPitchforks** [reply]
Why not follow Jan's suggestion in the original issue #36363? If you want to allow reflection, use a reflection type (`MethodInfo`) instead of raw pointers. Any interested party can always retrieve the function pointer from the reflection object (via `MethodInfo.MethodHandle.GetFunctionPointer()`). The consumer of the reflection object is now given a choice as to whether they want to use normal reflection APIs (`MethodInfo,Invoke`) or a fnptr-based system. Under this system there's no need for anybody to construct a delegate unless they want to.

Re: the source generator example, a source generator should have knowledge of the exact type of the fnptr, since it sits alongside the code that created that fnptr in the first place. Instead of passing around just the IntPtr, you could instead pass around a custom object that represents the (IntPtr, `typeof(delegate* managed<...>)`) tuple. Or use the existing `System.Reflection.Pointer` type for this, which is really just a glorified (void\*, Type) tuple. We'd need to add a `Pointer.Type` property getter (which we should probably do anyway), and we'd need to address https://github.com/dotnet/runtime/issues/11354 (which we're already considering). But the two of those combined should be able to convey the appropriate information without the need for `MethodInfo` if that route is viable.

**Reply by acaly** [reply]
Thanks for the suggestion. That's very similar to what I am thinking right now. I am really looking forward to the fix of #11354 (and probably dotnet/roslyn#48765) so that I can use this method to simplify it by replacing `MethodInfo` with a `typeof(delegate*<...>)`.

The justification of having such an API was really that, as @PathogenDavid mentioned, there is already a function in netcore that does exactly the same work, so making it publicly available to C# isn't that difficult, at least technically. It can be really useful with function pointer, to make the code more robust and to print more useful debug info. Note that without this API, even we can have the `typeof(delegate*<...>)`, we still need a delegate type to convert the pointer to `MethodInfo`. One may be able to construct such a delegate type as a (reusable) dynamic type if it is really necessary, but still it adds much complication to the user.

Do you think I should open another discussion or issue to discuss this? Or you think the team would choose not to implement this API? This discussion is becoming too long to read.
